### PR TITLE
New serverless pattern - Lambda response streaming from DynamoDB

### PR DESCRIPTION
**Pull request by sandykumar93**
*Issue #2482* 

*Description of changes:*
The newly added pattern shows how to use Lambda response streaming to incrementally retrieve and stream DynamoDB query / scan results using the write() method. Instead of waiting for the entire query / scan operation to complete, the Lambda function streams data in batches by setting a limit on the number of items per query and sending each batch as soon as it's retrieved. This improves the time-to-first-byte (TTFB) by streaming results to the client as they become available.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: lambda-streaming-ttfb-write-sam-with-dynamodb/README.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+The service interaction in this pattern uses AWS Lambda's response streaming capability to stream data in batches from Amazon DynamoDB. Instead of retrieving all query / scan results at once, the Lambda function processes the query incrementally, retrieving and sending results as they become available. Here are the details of the interaction:
+
+1. Client Request
+   A client sends a request to the Lambda URL, asking for specific data from a DynamoDB table.
+2. Lambda Function Initialization :
+   When the Lambda function is invoked, it initializes a connection to DynamoDB and prepares to run a query operation. The query includes a limit parameter, which restricts the number of items retrieved in each batch.
+3. Querying
+   The Lambda function begins querying DynamoDB using the defined limit (e.g., 100 items per batch). DynamoDB will return only a limited set of results instead of the entire result set at once.
+4. Response Streaming
+   As soon as a batch of results is retrieved, the Lambda function uses the `write()` method of Lambda's streaming API to send the data to the client. This happens immediately, without waiting for the entire query operation to complete.
```
Comment (ellisms) [inline lambda-streaming-ttfb-write-sam-with-dynamodb/README.md:63]:
```suggestion
   As soon as a batch of results is retrieved, the function uses Lambda's streaming API `write()` method to send the data to the client. This happens immediately, without waiting for the entire query operation to complete.
```

File: lambda-streaming-ttfb-write-sam-with-dynamodb/example-pattern.json [hunk 2]
```diff
@@ -0,0 +1,54 @@
+{
+  "title": "Lambda Response streaming: Streaming incremental DynamoDB Query results.",
+  "description": "This pattern uses Lambda response streaming to incrementally retrieve and stream DynamoDB results in batches, reducing time-to-first-byte (TTFB) by sending data as it's retrieved.",
```
Comment (ellisms) [inline lambda-streaming-ttfb-write-sam-with-dynamodb/example-pattern.json:3]:
```suggestion
  "description": "This pattern uses AWS Lambda response streaming to incrementally retrieve and stream Amazon DynamoDB results in batches, reducing time-to-first-byte (TTFB) by sending data as it's retrieved.",
```

**Review by ellisms — CHANGES_REQUESTED** [review]
Thanks for your submission @sandykumar93 . Requesting a few minor edits.

File: lambda-streaming-ttfb-write-sam-with-dynamodb/README.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
+# Lambda Response streaming: Streaming incremental DynamoDB Query results.
```
[the remark this answers is not in the document]
Comment (sandykumar93) [reply lambda-streaming-ttfb-write-sam-with-dynamodb/README.md]: Applied the suggested changes in the latest commit.

File: lambda-streaming-ttfb-write-sam-with-dynamodb/README.md [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+1. From the command line, use AWS SAM to deploy the AWS resources for the pattern as specified in the template.yml file:
+
+   ```
+   sam deploy -g --stack-name lambda-streaming-ttfb-write-sam-with-dynamodb
+   ```
+
+1. During the prompts:
+
+   - Enter a stack name
+   - Enter the desired AWS Region - AWS CLI default region is recommended if you are planning to run the test (test.sh) script
```
[the remark this answers is not in the document]
Comment (sandykumar93) [reply lambda-streaming-ttfb-write-sam-with-dynamodb/README.md]: Applied the suggested changes in the latest commit.

File: lambda-streaming-ttfb-write-sam-with-dynamodb/README.md (hunk 1 above)
Comment (sandykumar93) [reply lambda-streaming-ttfb-write-sam-with-dynamodb/README.md]: Applied the suggested changes in the latest commit.

File: lambda-streaming-ttfb-write-sam-with-dynamodb/example-pattern.json [hunk 5]
```diff
@@ -0,0 +1,54 @@
+{
+  "title": "Lambda Response streaming: Streaming incremental DynamoDB Query results.",
```
[the remark this answers is not in the document]
Comment (sandykumar93) [reply lambda-streaming-ttfb-write-sam-with-dynamodb/example-pattern.json]: Applied the suggested changes in the latest commit.

File: lambda-streaming-ttfb-write-sam-with-dynamodb/example-pattern.json (hunk 2 above)
Comment (sandykumar93) [reply lambda-streaming-ttfb-write-sam-with-dynamodb/example-pattern.json]: Applied the changes in the latest commit

File: lambda-streaming-ttfb-write-sam-with-dynamodb/README.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+## Testing
+
+1. Use curl with your AWS credentials to view the streaming response as the url uses AWS Identity and Access Management (IAM) for authorization. Replace the URL and Region parameters for your deployment.
+
+   ```
+   curl --request GET https://<url>.lambda-url.<Region>.on.aws/ --user AKIAIOSFODNN7EXAMPLE:wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY --aws-sigv4 'aws:amz:<Region>:lambda'
+   ```
+
+You can see the gradual display of the streamed response.
```
[the remark this answers is not in the document]
Comment (sandykumar93) [reply lambda-streaming-ttfb-write-sam-with-dynamodb/README.md]: @ellisms 

To address this, I have added a new lambda, updated the template and readme accordingly. The data source used in the code is a subset of [Spotify Dataset](https://www.kaggle.com/datasets/yamaerenay/spotify-dataset-19212020-600k-tracks?select=tracks.csv).

**Comment by sandykumar93** [comment]
@ellisms thank you for taking the time to review and provide suggestions. I have applied all the suggestions in the latest commit. Please have a look when possible and let me know if any other changes are required

File: lambda-streaming-ttfb-write-sam-with-dynamodb/README.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+   sam deploy -g --stack-name lambda-streaming-ttfb-write-sam-with-dynamodb
+   ```
+
+1. During the prompts:
+
+   - Enter a stack name
+   - Enter the desired AWS Region
+   - Allow SAM CLI to create IAM roles with the required permissions.
+
+1. Once you have run `sam deploy --guided` mode once and saved arguments to a configuration file `samconfig.toml`, you can use `sam deploy` in future to use these defaults.
```
Comment (ellisms) [inline lambda-streaming-ttfb-write-sam-with-dynamodb/README.md:42]:
```suggestion
1. After running `sam deploy --guided` mode once and saving arguments to a configuration file `samconfig.toml`, you can use `sam deploy` in future to use these defaults.
```

**Review by ellisms — CHANGES_REQUESTED** [review]
Just one last edit.

**Review by ellisms — APPROVED** [review]

**Comment by ellisms** [comment]
Thanks, adding to the publishing queue.

**Pull request merged**